Handle hyphen in @param doc comment

connorjs opened the pull request:
### ℹ️ Overview

Fixes [#2390](https://github.com/microsoft/typespec/issues/2390).

Strips the leading hyphen in `@param` doc comments. Preserves handling `@param` without the hyphen.

Parser change: The hyphen in the following code now tokenizes as `Token.Hyphen` instead of `Token.DocText`.

```
@param foo - A param doc with a leading hyphen
```

### 📝 Notes

I made the change in the parser so that the doc comment modeling only sees the string. I used an `if` check for `"param"` that maybe could be handled differently. I’m open to any feedback (helper function, etc), or just to hand this off to y’all to close out at your convenience.

### 🧪 Testing

Modified an existing unit test that happened to have two `@param` usages already. Let me know if we should add other unit tests.

connorjs commented:
@microsoft-github-policy-service agree

timotheeguerin commented on packages/compiler/src/core/parser.ts:
```diff
@@ -2425,7 +2425,20 @@ function createParser(code: string | SourceFile, options: ParseOptions = {}): Pa
+    if (messageId === "param") {
+      // The TSDoc spec expects a hyphen in the `@param` case, but it does
+      // not require it in practice. Thus, TypeSpec should handle both cases.
+      while (parseOptional(Token.Whitespace));
+      if (token() === Token.Hyphen) {
+        // The doc content starts with a hyphen, so skip it and subsequent whitespace
+        nextToken();
```
```suggestion
      while (parseOptional(Token.Whitespace));
      if (parseOptional(Token.Hyphen)) {
        // The doc content starts with a hyphen, so skip it and subsequent whitespace
```
I think you can use  `if(parseOptional(Token.Hyphen)` instead and skip calling nextToken below

timotheeguerin commented on packages/compiler/src/core/parser.ts:
```diff
@@ -2425,7 +2425,20 @@ function createParser(code: string | SourceFile, options: ParseOptions = {}): Pa
   ): ParamLikeTag {
     const name = parseDocIdentifier(messageId);
+
+    if (messageId === "param") {
```
This feels like a little hacky of a check, comparing on a message code.

is `@param` the only one that allows `-` in tsdoc? I think I would just remove this check and allow it for all `ParamLikeTag` that this function is parsing.

timotheeguerin replied:
Looking at `@typeParam` https://tsdoc.org/pages/tags/typeparam/ seems to be also using hyphen so I would just remove that if and always try to parse `-`

connorjs replied:
Oh! 🤦🏻‍♂️ `typeparam` = `@typeParam`. I had not realized that; yes, I can re-evaluate and share this logic.

I will update today or tomorrow.

timotheeguerin replied:
We have `@template` which is `@typeParam` in tsdoc(or also allows `@template`) and `@param` (same as in tsdoc) both of those are `ParamLikeTag` which have a parameter name and I believe then we can argue could have this optional `-`

timotheeguerin reviewed (commented):
Thanks a lot for the contribution this looks good! We'll most likely need to go through this in a design meeting to make sure we want to allow this. (As well as discuss your suggestion on making `-` the recommended syntax)

connorjs commented:
Thanks @timotheeguerin ! I agree with confirming this with the team before merging (plus the recommendation side of things). No rush on my side (easy workaround: no hyphen); I’m happy to help as I can..

connorjs commented:
@timotheeguerin  - I’ve updated the implementation to handle all param-like tags. I added some function-level doc comments and created a new helper function - feel free to trim or edit any of that as you see fit.

I also added test cases for the template params.

(And yep - still pending the team discussion I expect 👍🏻)

Petermarcu commented:
@timotheeguerin any update here?

timotheeguerin commented:
@Petermarcu we just need to approve this minor syntax addition in a design meeting, otherwise the implementation looks good minus some failures in the CI

connorjs commented:
Let me know if any of the CI build failures need work on my side. Thanks for keeping a watch on this y’all.

timotheeguerin commented:
@connorjs it looks like the CI is failing on your new tests

connorjs commented:
Thanks @timotheeguerin - let me update my local branches and check this out. I’ll get an update published before y’all’s Thursday design meeting. 👍🏻

connorjs commented:
I have removed the failing tests because the existing compiler code does not preserve template doc comments.

On second look, this makes sense to me because the template parameters **only** apply at library/build time. They have no value in the final TypeSpec (per my understanding).

For example, no Open API (or similar) finalized/generated API document would include the templates. I think.

Question: Is my reasoning correct? Or should TypeSpec preserve these? For example, would code generators want to preserve these templated types and generate them as the intermediate types? (FWIW, I argue against preserving the intermediate types b/c I find they exist for authors, not consumers.)

Side note: I printed all doc comments and that’s how I observed the existing `@template` doc was not preserved on the compile stage. (The parser does have the doc comments.)

Here's a piece of my hacky println debugging: `[...getAllDocData(runner.program)?.values() ?? []].map(v=>"value" in v ? v.value : v)`

_If that doesn’t make sense, I will type out a more composed message tomorrow._

timotheeguerin commented:
You are correct that the `@template` wouldn't produce doc accessible with getDoc or similar however you can still read the doc manually after the parsing. It might be good to add a test there to make sure the doc node doesn't contain the hyphen.

connorjs commented:
>It might be good to add a test there to make sure the doc node doesn't contain the hyphen.

I have added that for the `@param`, but not `@template` (and `@typeParam`?).

I can take another look tonight at the `@template` compiler logic. (The issue I was running into.)

timotheeguerin commented:
yep meant to add test for the parser so you don't call `getDoc` which won't have the value of `@template`

Example on the ref doc of how to read the nodes https://github.com/microsoft/typespec/blob/b9452d0ffeb1e2f7d84c6f92b63ff26466840eeb/packages/tspd/src/ref-doc/extractor.ts#L521

timotheeguerin commented on packages/compiler/src/core/parser.ts:
```diff
@@ -2418,14 +2418,23 @@ function createParser(code: string | SourceFile, options: ParseOptions = {}): Pa
+   * Handles param-like documentation comment tags: `@param` and
+   * `@typeParam` (or `@template`).
+   *
+   * @see <a href="https://tsdoc.org/pages/tags/param/">@param</a>
```
Not sure we should link to the typescript docs for this. 

I think instead we should make sure all the tags are documented and linkable in this doc https://microsoft.github.io/typespec/language-basics/documentation#tsdoc-doc-comments

timotheeguerin replied:
for this one you can maybe just link the docs page we can have a follow up to update the doc itself

connorjs replied:
Reorganized/wrote the doc comments. 👍🏻

timotheeguerin commented on packages/compiler/src/core/parser.ts:
```diff
@@ -2435,6 +2444,21 @@ function createParser(code: string | SourceFile, options: ParseOptions = {}): Pa
+   * The TSDoc spec expects a hyphen in the `@param` case, but it does not
+   * require it in practice. Thus, TypeSpec should handle both cases.
+   */
+  function parseParamLikeHyphen() {
```
```suggestion
  function parseOptionalParamLikeHyphen() {
```

connorjs replied:
Thanks, with my rewording I made it `parseOptionalHyphenDocParamLikeTag` for parallel naming.

connorjs commented:
Thanks for the comments and recommendations. I will update tomorrow (along with the tests already mentioned). Hopefully we can get merged by Monday to meet the October release 🤞🏻

connorjs commented:
@timotheeguerin - Happy October! Let me know if y’all need anything from me for this PR.

timotheeguerin reviewed (approved):
Awesome thanks a lot for the contribution and your patience!
